include/openmc/shared_array.h

```diff
@@ -0,0 +1,119 @@
+#ifndef OPENMC_SHARED_ARRAY_H
+#define OPENMC_SHARED_ARRAY_H
+
+//! \file shared_array.h
+//! \brief Shared array data structure
+
+#include<memory>
+
+
+namespace openmc {
+
+//==============================================================================
+// Class declarations
+//==============================================================================
+
+// The SharedArray is an array that is capable of being appended to in an
+// thread safe manner by use of atomics. It only provides protection for the
+// use cases currently present in OpenMC. Namely, it covers the scenario where
+// multiple threads are appending to an array, but no threads are reading from
+// or operating on it in any other way at the same time. Multiple threads can
+// call the thread_safe_append() function concurrently and store data to the
+// object at the index returned from thread_safe_append() safely, but no other
+// operations are protected.
+template <typename T> 
+class SharedArray { 
+
+private: 
+  //==========================================================================
+  // Data members
+
+  std::unique_ptr<T[]> data_; //!< A pointer to hold the data
+  int64_t size_ {0}; //!< The current size of the SharedArray. 
+  int64_t capacity_ {0}; //!< The maximum capacity of the SharedArray.
```

Comment:
These comments don't add any information. May as well remove them.

Reply:
If we are going to make this class "STL-like", perhaps we should have a `size_type` (which will default to `std::size_t`) and use it for these data members as well as for function arguments where we're currently taking an int64_t.

Reply:
I see two downsides to adding a `size_type` member:

1 - Currently, the `thread_safe_append()` function returns a -1 if the container is full. So, if we wanted to switch this to a `size_t` we would need to start passing in an error code variable as an extra reference argument to this function. Or, we could leave that return index as a `int64_t`, but change the rest to a `size_type`, which is fine but doesn't seem all the elegant to me.

2 - The class only provides a small subset of functionality that we might typically expect in a STL class, so I don't think we need to necessarily use all of the conventions that a more generic class would use. As we are not really implementing a truly generic class, and as I can't conceive of a case in which a `int64_t` would not be appropriate for our use cases in OpenMC, this seems like it just adds another member and complicates the code without tangible benefit.

Reply:
In response to @smharper's comment, I added the comments per @paulromano's direction in a previous PR in which [he said](https://github.com/openmc-dev/openmc/pull/1451#discussion_r367105412) 

> It's pretty obvious, but good practice to put some trailing `//!<` comments describing the members

Reply:
Even better would be to modify the comments so they explain/hint at the differences between `size` and `capacity`. The name `data` is pretty self-explanatory, but the comment can indicate the motivation for a unique pointer. Eg:
```C++
  std::unique_ptr<T[]> data_; //!< An RAII handle to the underlying data
  int64_t size_ {0}; //!< The current number of elements
  int64_t capacity_ {0}; //!< The total space allocated for elements
```
(Plus, it's better to avoid sticking the name `SharedArray` in the comments to prevent errors in case someone changes the name of the data structure in the future or copy/pastes the code elsewhere.)

Reply:
Sounds good, will do.